client/mysqltest.cc

```diff
@@ -3865,26 +3865,268 @@ void do_remove_files_wildcard(struct st_command *command)
+  size_t dest_file_length;
+  dest_file_length= ds_destination->length;
+  source_file_length= ds_source->length;
+  uint match_count;
+  match_count= 0;
+
+  for (uint i= 0; i < src_dir_info->number_of_files; i++)
+  {
+    ds_source->length= source_file_length;
+    ds_destination->length= dest_file_length;
+    FILEINFO *file= src_dir_info->dir_entry + i;
+
+    /* Skip directories if recursive copy is not requested */
+    if (MY_S_ISDIR(file->mystat->st_mode) && !recursive)
+      continue;
+
+    /* Copy only those files which the pattern matches */
+    if (ds_wild->length &&
+        wild_compare(file->name, ds_wild->str, false))
+      continue;
+
+    match_count++;
+    dynstr_append_mem(ds_source, file->name, strlen(file->name));
+    dynstr_append_mem(ds_destination, file->name, strlen(file->name));
+
+    if (MY_S_ISDIR(file->mystat->st_mode))
+    {
+      DBUG_PRINT("info", ("Recursive copy files of %s to %s",
+                          ds_source->str, ds_destination->str));
+      DBUG_PRINT("info", ("listing directory: %s", ds_source->str));
+      error= recursive_copy(ds_source, ds_destination, do_not_overwrite, false);
+    }
+    else
+    {
+      DBUG_PRINT("info", ("Copying file: %s to %s",
+                          ds_source->str, ds_destination->str));
+      /* MY_HOLD_ORIGINAL_MODES prevents attempts to chown the file */
+      error= (my_copy(ds_source->str, ds_destination->str,
+              MYF(MY_HOLD_ORIGINAL_MODES)) != 0);
+    }
+
+    if (error) goto end;
+  }
+
+  /* Pattern did not match any files */
+  if (!match_count)
+  {
+    error= 1;
+  }
+
+end:
+  my_dirend(src_dir_info);
+  my_dirend(dst_dir_info);
+  return error;
+}
+
+
 /*
   SYNOPSIS
   do_copy_file
   command	command handle
 
   DESCRIPTION
-  copy_file <from_file> <to_file>
-  Copy <from_file> to <to_file>
+  copy_file <source> <destination> [-r|-f|-rf]
+
+  <source> can be a file, a directory, or a path with wildcard pattern
 
-  NOTE! Will fail if <to_file> exists
+  - When <source> is a directory, '-r' must be given to copy recursively
+  - When <source> is a wildcard pattern, '-r' is optional. If given, then
+    directories that match the pattern will also be copied.
+  - When '-f' is given, files in destination with same name will be overwritten
+  - Wildcard is only supported in the basename of path
+    (e.g. src/(wildcard)/(wildcard) will not work)
 */
 
 void do_copy_file(struct st_command *command)
 {
   int error;
-  static DYNAMIC_STRING ds_from_file;
-  static DYNAMIC_STRING ds_to_file;
-  const struct command_arg copy_file_args[] = {
-    { "from_file", ARG_STRING, TRUE, &ds_from_file, "Filename to copy from" },
-    { "to_file", ARG_STRING, TRUE, &ds_to_file, "Filename to copy to" }
+  static DYNAMIC_STRING ds_source;
+  static DYNAMIC_STRING ds_destination;
+  static DYNAMIC_STRING ds_wild;
+  static DYNAMIC_STRING ds_options;
+  static bool recursive;
+  static int do_not_overwrite; // Flags for my_copy()
```

Comment:
Why are all these static?
ds_source/ds_destination/ds_options - I can understand that, at least there was a reason for that twenty years ago. But others?